client/landing/stepper/declarative-flow/internals/steps-repository/woo-transfer/index.tsx

```diff
@@ -0,0 +1,114 @@
+/* eslint-disable no-console */
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect } from 'react';
+import { useSite } from 'calypso/landing/stepper/hooks/use-site';
+import { transferStates } from 'calypso/state/atomic/transfers/constants';
+import { ONBOARD_STORE, SITE_STORE } from '../../../../stores';
+import type { Step } from '../../types';
+
+const wait = ( ms: number ) => new Promise( ( res ) => setTimeout( res, ms ) );
+
+const WooTransfer: Step = function WooTransfer( { navigation } ) {
+	const { submit } = navigation;
+	const { setPendingAction, setProgress } = useDispatch( ONBOARD_STORE );
+	const {
+		requestAtomicSoftwareStatus,
+		requestLatestAtomicTransfer,
+		initiateAtomicTransfer,
+	} = useDispatch( SITE_STORE );
+	const site = useSite();
+
+	const siteId = site?.ID;
+
+	const {
+		getSiteLatestAtomicTransfer,
+		getSiteLatestAtomicTransferError,
+		getAtomicSoftwareStatus,
+		// getAtomicSoftwareError,
+	} = useSelect( ( select ) => select( SITE_STORE ) );
+
+	useEffect( () => {
+		if ( ! siteId ) return;
+
+		setPendingAction( async () => {
+			const startTime = new Date().getTime();
+			const totalTimeout = 1000 * 180;
+			const maxFinishTime = startTime + totalTimeout;
+
+			// Initiate transfer
+			await initiateAtomicTransfer( siteId, 'woo-on-plans' );
+
+			// Poll for transfer status
+			let stopPollingTransfer = false;
+
+			while ( ! stopPollingTransfer ) {
+				await wait( 3000 );
+				await requestLatestAtomicTransfer( siteId );
+				const transfer = getSiteLatestAtomicTransfer( siteId );
+				const transferError = getSiteLatestAtomicTransferError( siteId );
+				const transferStatus = transfer?.status;
+				const isTransferringStatusFailed = transferError && transferError?.status >= 500;
+
+				switch ( transferStatus ) {
+					case transferStates.PENDING:
+						setProgress( 0.2 );
+						break;
+					case transferStates.ACTIVE:
+						setProgress( 0.4 );
+						break;
+					case transferStates.PROVISIONED:
+						setProgress( 0.5 );
+						break;
+					case transferStates.COMPLETED:
+						setProgress( 0.7 );
+						break;
+				}
+
+				if ( isTransferringStatusFailed || transferStatus === transferStates.ERROR ) {
+					throw new Error( 'Transfer error' );
+					// onFailure( {
+					// 	type: 'transfer',
+					// 	error: transferError?.message || softwareError?.message || '',
+					// 	code: transferError?.code || softwareError?.code || '',
+					// } );
+				}
+
+				if ( maxFinishTime < new Date().getTime() ) {
+					throw new Error( 'Timeout error' );
+				}
+
+				stopPollingTransfer = transferStatus === transferStates.COMPLETED;
+			}
+
+			// Poll for software status
+			let stopPollingSoftware = false;
+
+			while ( ! stopPollingSoftware ) {
+				await wait( 3000 );
```

Comment:
The software installation is requested in the `initiateAtomicTransfer` action, so in this case, it might be better to push the `wait` to the bottom of the loop.

Reply:
Cool! I've also make the waiting conditional in this case, so if it's ready it doesn't wait for an extra 3s